src/serializer/ResidualHeapSerializer.js

```diff
@@ -1229,6 +1229,22 @@ export class ResidualHeapSerializer {
     return serializedValue;
   }
 
+  _serializeAbstractValue(val: AbstractValue): void | BabelNodeExpression {
+    invariant(val.kind !== "sentinel member expression", "invariant established by visitor");
+    if (val.hasIdentifier()) {
+      return this._serializeAbstractValueHelper(val);
+    } else {
+      // This abstract value's dependencies may have been declared
+      // but still need to check it again in case their serialized bodies are in different generator scope.
+      this.emitter.emitNowOrAfterWaitingForDependencies(val.args, () => {
+        const serializedValue = this._serializeAbstractValueHelper(val);
+        let uid = this.residualHeapValueIdentifiers.getIdentifierAndIncrementReferenceCount(val);
+        let declar = t.variableDeclaration("var", [t.variableDeclarator(uid, serializedValue)]);
+        this.emitter.emit(declar);
```

Comment:
I am confused about this variable declaration. Doesn't that compete with another declaration? This deserves a comment (and I'd like to understand it).

Reply:
This code just serializes the abstract value into "serializedValue", then add a var declaration for it:
```js
var uid = < serializedValue>;
```
Finally it just calls emitter to emit it.
For "Doesn't that compete with another declaration", what is "another declaration"?

Reply:
But you are not really making up a new identifier, you just get the identifier assigned to that value anyway. So there should already be a declaration somewhere just like the one you are emitting.

Reply:
I see your question now. 
There is not another declaration. Before _serializeAbstractValue is called(actually before any _serializeValue call), this.residualHeapValueIdentifiers.setIdentifier() is called to assign a new identifier for a Value.(This is done in serializeValue() method).
So residualHeapValueIdentifiers.getIdentifierAndIncrementReferenceCount() can always fetch the previous assigned identifier.
I will add a comment on "getIdentifierAndIncrementReferenceCount()" call with above explanation. But I am not sure if it is necessary because I followed the pattern of other places in this file using emitNowOrAfterWaitingForDependencies(): most of them call getIdentifierAndIncrementReferenceCount() to fetch the previously assigned identifier with the same assumption.